WISE-Community/WISE-API
# feat(PeerGrouping): Pair students with different KI scores

## Changes
- Add ```differentKIScores(nodeId: string, componentId: string, mode?: 'any' | 'maximize')``` term in PeerGrouping.logic that pairs students based on different KI scores
- Refactor code. Extract common code to parent (abstract) classes.

## Test Prep
1. Create a CRater component with a KI score e.g. "MusicalInstruments_Speed_score-ki_idea-ki_nonscor"
2. Add 2 new PeerGroupings to unit content (via JSON editing), and set the logic's nodeId and componentId to point to to the new component.
```
"peerGroupings": [
...
  {
    "logic": "differentKIScores(\"NODE_ID\",\"COMPONENT_ID\",\"any\")",
    "maxMembershipCount": 2,
    "name": "Different KI Score Any",
    "tag": "dg0uytsun8"
  },
  {
    "logic": "differentKIScores(\"NODE_ID\",\"COMPONENT_ID\",\"maximize\")",
    "maxMembershipCount": 2,
    "name": "Different KI Score Maximize",
    "tag": "yuzhceiwfho"
  }
...
]
``` 
3. Save these in the peer_groupings table if the unit is used in a run. Saving part doesn't work if you edit the JSON directly.
4. Create 2 new components that use PeerGroup (e.g. PeerChat, Show Group Work):
   - One should specify "Different KI Score Any" as its Grouping Logic ("PeerGroupComponent1")
   - Another should specify "Different KI Score Maximize" as its Grouping Logic ("PeerGroupComponent2")

## Test DifferentKIScore Logic with "any" mode
- When the student gets to PeerGroupComponent1, they should be paired with another workgroup if their KI scores are different (it doesn't matter by how much). 
- Student should only paired with another workgroup with the same KI score if that's the only choice left.

```
Workgroup: KIScore
Workgroup1: 3
Workgroup2: 2
Workgroup3: 3
Workgroup4: 4
Workgroup5: 5
```

In this case, Workgroup1 should be paired with Workgroup2, Workgroup4, or Workgroup5. 

## Test DifferentKIScore Logic with "maximize" mode
- When the student gets to PeerGroupComponent2, they should be paired with another workgroup if their KI scores are different, and it should be with the one with the biggest difference. 
- Again, student should only paired with another workgroup with the same KI score if that's the only choice left.

In the same case as above, Workgroup1 should be paired with Workgroup5 since they have difference of 2 in the KI scores.

## Test Random and DifferentIdeas pairing
- Random and DifferentIdeas pairing should work as before. The related code has been refactored, but no new feature was added as far as these logic are concerned.

Closes #184
